confmap/confmap.go

```diff
@@ -431,6 +432,40 @@ func unmarshalerEmbeddedStructsHookFunc() mapstructure.DecodeHookFuncValue {
 	}
 }
 
+// optionalHookFunc applies logic when the to.Type is optional.Optional. When decoding primitive types, 
+// we use reflect.Value's FieldByName API (https://pkg.go.dev/reflect#Value.FieldByName) and Set (https://pkg.go.dev/reflect#Value.Set) APIs
+// in order to set the Value and HasValue. Set cannot be used for struct field types, as the set value must be assignable to the target type, 
+// so for structs we use optionals custom unmarshaller in order to cast the map to the struct.
+//
+// This logic relies on the fact that the hook only gets called when the value is explicitely set in the config,
+// so if we are in this hook, we know we can set HasValue to true.
+func optionalHookFunc() mapstructure.DecodeHookFuncValue {
+	return func(from reflect.Value, to reflect.Value) (any, error) {
+		// As the type is generic, we check for the prefix.
+		if strings.HasPrefix(to.Type().String(), "optional.Optional") {
+			// the optional.Optional field is a struct
+			if _, ok := from.Interface().(map[string]any); ok {
```

Comment:
Could we allow this block to be handled by `unmarshalerHookFunc`?

Reply:
Good point, removed this logic